src/server/engine_shard_set.cc

```diff
@@ -72,37 +88,44 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+void EngineShard::DefragTaskState::UpdateState(uint64_t cursor_val) {
+  cursor = cursor_val;
+  if (cursor == kCursorDoneState) {
+    // This means that the scan in not active, so we
+    // don't know what the memory state any more, and we need
+    // to try and detect memory utilization again
+    underutilized_found = false;
+  }
+}
+
 // This function checks 3 things:
 // 1. Don't try memory fragmentation if we don't use "enough" memory (control by
 // mem_defrag_threshold flag)
 // 2. That we had change in memory usage - to prevent endless loop - out of scope for now
 // 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
 // (control by commit_use_threshold flag)
-bool EngineShard::DefragTaskState::IsRequired() const {
-  const int64_t threshold_mem = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
-  const double commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
-
-  if (cursor > kCursorDoneState) {
+bool EngineShard::DefragTaskState::CheckRequired() {
+  if (cursor > kCursorDoneState || underutilized_found) {
+    VLOG(1) << "Already found memory utilization issue - cursor: " << cursor
+            << " and underutilized_found " << underutilized_found;
     return true;
   }
+  const int64_t threshold_mem = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
+  const double commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
 
   // can be negative due to weird accounting of mimalloc.
   int64_t commited = GetMallocCurrentCommitted();
 
-  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
```

Comment:
we should not use `GetMallocCurrentCommitted`

Reply:
it was removed here

Reply:
I will remove it, I think that it is not required any more.